### PR TITLE
auth LUA: add dblookup function

### PR DESCRIPTION
**Pull request by Habbie**
### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Comment by Habbie** [comment]
zeha: depending on usage context, the user may want to put `[1]` after dblookup to get one item instead of a list.

dwfreed: "Some things, like ifurlup, don't like empty tables, so be careful not to accidentally look up a name that does not have any records of that type"

**Comment by Habbie** [comment]
after squash, commit message should mention that this fixes #12358

File: pdns/lua-record.cc [hunk 1]
```diff
@@ -1103,6 +1115,36 @@ static void setupLuaRecords(LuaContext& lua)
       return result;
     });
 
+  lua.writeFunction("dblookup", [](const string& record, const string& type) {
+    DNSName rec;
+    QType qtype;
+    vector<string> ret;
+    try {
+      rec = DNSName(record);
+      qtype = type;
```
Comment (omoerbeek) [inline pdns/lua-record.cc:1124]: `QType &operator=(const string &)` does not throw an exception on error, but assigns zero.

File: pdns/lua-record.cc [hunk 2]
```diff
@@ -1103,6 +1115,36 @@ static void setupLuaRecords(LuaContext& lua)
...
+    }
+    catch (const std::exception& e) {
+      g_log << Logger::Error << "DB lookup cannot be performed, the name (" << record << ") or type (" << type << ") is malformed: " << e.what() << endl;
```
Comment (omoerbeek) [inline pdns/lua-record.cc:1127]:
```suggestion
      g_log << Logger::Error << "DB lookup cannot be performed, the name (" << record << ") or type (" << qtype << ") is malformed: " << e.what() << endl;
```

File: pdns/lua-record.cc [hunk 3]
```diff
@@ -1103,6 +1115,36 @@ static void setupLuaRecords(LuaContext& lua)
...
+        return ret;
+      }
+
+      vector<DNSZoneRecord> drs = lookup(rec, qtype, soaData.domain_id);
+      for (const auto& drec : drs) {
+        ret.push_back(drec.dr.getContent()->getZoneRepresentation());
+      }
+    }
+    catch (std::exception& e) {
+      g_log << Logger::Error << "Failed to do DB lookup for " << rec << "/" << type << ": " << e.what() << endl;
```
Comment (omoerbeek) [inline pdns/lua-record.cc:1143]:
```suggestion
      g_log << Logger::Error << "Failed to do DB lookup for " << rec << "/" << qtype << ": " << e.what() << endl;
```

File: pdns/lua-record.cc [hunk 4]
```diff
@@ -503,6 +504,17 @@ static std::vector<DNSZoneRecord> lookup(const DNSName& name, uint16_t qtype, in
   return ret;
 }
 
+static bool getAuth(const DNSName& name, uint16_t qtype, SOAData* soaData)
+{
+  static LockGuarded<UeberBackend> s_ub;
+
+  vector<DNSZoneRecord> ret;
```
Comment (omoerbeek) [inline pdns/lua-record.cc:511]: Looks unused

File: pdns/lua-record.cc (hunk 1 above)
Comment (Habbie) [reply pdns/lua-record.cc:1124]: good catch, will handle

File: pdns/lua-record.cc (hunk 2 above)
Comment (Habbie) [reply pdns/lua-record.cc:1127]: hmm, if we reach this line, we know for a fact qtype has not been set, right?

Comment (omoerbeek) [reply pdns/lua-record.cc:1127]: Hmm, right i mixed the types up.

File: pdns/lua-record.cc (hunk 1 above)
Comment (Habbie) [reply pdns/lua-record.cc:1124]:
```suggestion
      qtype = type;
      if (qtype.getCode() == 0) {
        throw std::invalid_argument("unknown type");
      }
```

plus leave `type` in the next log line?

Comment (omoerbeek) [reply pdns/lua-record.cc:1124]: yes

File: pdns/lua-record.cc (hunk 3 above)
Comment (Habbie) [reply pdns/lua-record.cc:1143]: i did apply this one

**Comment by Habbie** [comment]
thank you, should all be resolved now

**Pull request merged**